### DISCUSSION TITLE
Optimization: chakra factory or custom defined component or variant/layerStyle

### DISCUSSION THREAD

**Discussion by willzeng274** [chakra-ui/chakra-ui#5884] [Q&A]
Hey guys, I'm really confused about which one to use to reduce bundle size and runtime for clients.

Chakra UI's factory with baseStyle, a custom-defined React component or use a variant/layerStyle for that component.

These are the options I'm talking about:

```ts
// factory with baseStyle
const Card = chakra(Box, {
  baseStyle: {
    shadow: 'lg',
    rounded: 'lg',
    bg: 'white',
  },
});

// custom-defined React component
const Card = ({ children, ...props }) => (
  <Box
    shadow="lg"
    rounded="lg"
    bg="white"
    {...props}
  >
    {children}
  </Box>
);

// using layerStyle in extendTheme
const theme = extendTheme({
  layerStyles: {
    card1: {
      shadow: 'lg',
      rounded: 'lg',
      bg: 'white',
    }
  },
});
```
Though, I think useColorModeValue is not available for use in the chakra factory and extendTheme (?)

I rely heavily on breakpoint values and color mode values, but sometimes I do have pure style components.

Does anyone have any suggestions on when to use these methods?

Thanks in advance

(Note: my main goal is to speed up my app and reduce bundle size, this can include reusing styles! If that reduces bundle size)

**Comment by hc0503** [comment]
This is very important discussion for high performance application development.
We can use color mode in extendTheme.
This is my code.

```ts
const Input: ComponentStyleConfig = {
  variants: {
    primary: (props) => ({
      field: {
        fontSize: "0.8125rem",
        fontWeight: "semibold",
        borderWidth: "2px",
        borderRadius: "0.25rem",
        borderColor: "transparent",
        bgColor: mode("gray.990", "green.700")(props),
        height: "unset",
        py: "0.375rem",
        px: "0.75rem",
        _focus: "green.300",
        _hover: "green.300",
        _invalid: {
          borderColor: "red.700",
        },
      },
    }),
  },
};
```
